kernel/kernel-api/src/main/java/io/delta/kernel/internal/snapshot/SnapshotManager.java

```diff
@@ -156,22 +157,26 @@ public static void verifyDeltaVersionsContiguous(List<Long> versions, Path table
 
   private SnapshotImpl createSnapshot(
       LogSegment initSegment, Engine engine, SnapshotQueryContext snapshotContext) {
-    // Note: LogReplay now loads the protocol and metadata (P & M) only when invoked (as opposed to
-    //       eagerly in its constructor). Nonetheless, we invoke it right away, so SnapshotImpl is
-    //       still constructed with an "eagerly"-loaded P & M.
+    final ProtocolMetadataLogReplay.Result protocolMetadataResult =
+        ProtocolMetadataLogReplay.loadProtocolAndMetadata(
+            engine, initSegment, tablePath, snapshotContext.getSnapshotMetrics());
 
     final LogReplay logReplay =
         new LogReplay(
-            tablePath, engine, new Lazy<>(() -> initSegment), snapshotContext.getSnapshotMetrics());
+            tablePath,
+            engine,
+            new Lazy<>(() -> initSegment),
+            protocolMetadataResult.crcInfoUsed,
+            snapshotContext.getSnapshotMetrics());
 
     final SnapshotImpl snapshot =
         new SnapshotImpl(
             tablePath,
             initSegment.getVersion(),
             new Lazy<>(() -> initSegment),
             logReplay,
-            logReplay.getProtocol(),
-            logReplay.getMetadata(),
```

Comment:
Before, it was weird that we (1) load P & M and then (2) inject a LogReplay and those P & M into Snapshot, even though LogReplay still has the `getProtocol, getMetadata` APIs. Basically, we are injecting a thing into Snapshot that can give you the P & M, but we already loaded the P & M when we creatd the snapshot!